cmd/posix.go

```diff
@@ -541,68 +541,139 @@ func (s *posix) ReadFile(volume string, path string, offset int64, buf []byte) (
 	return int64(m), err
 }
 
-// AppendFile - append a byte array at path, if file doesn't exist at
-// path this call explicitly creates it.
-func (s *posix) AppendFile(volume, path string, buf []byte) (err error) {
+func (s *posix) createFile(volume, path string) (f *os.File, err error) {
 	defer func() {
 		if err == syscall.EIO {
 			atomic.AddInt32(&s.ioErrCount, 1)
 		}
 	}()
 
 	if s.ioErrCount > maxAllowedIOError {
-		return errFaultyDisk
+		return nil, errFaultyDisk
 	}
 
 	// Validate if disk is free.
 	if err = s.checkDiskFree(); err != nil {
-		return err
+		return nil, err
 	}
 
 	volumeDir, err := s.getVolDir(volume)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	// Stat a volume entry.
 	_, err = os.Stat(preparePath(volumeDir))
 	if err != nil {
 		if os.IsNotExist(err) {
-			return errVolumeNotFound
+			return nil, errVolumeNotFound
 		}
-		return err
+		return nil, err
 	}
+
 	filePath := pathJoin(volumeDir, path)
 	if err = checkPathLength(filePath); err != nil {
-		return err
+		return nil, err
 	}
+
 	// Verify if the file already exists and is not of regular type.
 	var st os.FileInfo
 	if st, err = os.Stat(preparePath(filePath)); err == nil {
 		if !st.Mode().IsRegular() {
-			return errIsNotRegular
+			return nil, errIsNotRegular
 		}
 	}
```

Comment:
minor optimization, you can put and else {...} statement here and do mkdirAll in the else statement.
this optimization is not related to the current change set.